src/cstringholder.rs

```diff
@@ -0,0 +1,173 @@
+use std::ptr;
+use translate::*;
+use types::{StaticType, Type};
+
+use gobject_ffi;
+use ffi as glib_ffi;
+use value::{FromValueOptional, Value};
+
+#[derive(Debug, Hash, PartialOrd, Ord)]
+pub struct CStringHolder {
+    slice: Box<CStr>,
+    ptr: *const c_char,
+    owned: bool,
+}
+
+impl CStringHolder {
+    pub fn new(ptr: *const c_char) -> Self {
+        assert!(!ptr.is_null());
+        unsafe {
+            let slice = CStr::from_ptr(ptr);
+            Self { ptr, slice: Box::from(slice), owned: true }
+        }
+    }
+
+    pub fn from(s: &[u8]) -> Self {
+        let slice = match CStr::from_bytes_with_nul(s) {
+            Ok(s) => s,
+            Err(e) => panic!(e),
+        };
+        let ptr = slice.as_ptr();
+        assert!(!ptr.is_null());
+        Self { ptr, slice: Box::from(slice), owned: false }
+    }
+
+    pub fn as_str(&self) -> &str {
+        let bytes = self.slice.to_bytes();
+        match std::str::from_utf8(bytes) {
+            Ok(s) => s,
+            Err(e) => {
+                panic!("UTF-8 conversion failed: {}", e);
+            },
+        }
+    }
+}
+
+impl Drop for CStringHolder {
+    fn drop(&mut self) {
+        if self.owned {
+            unsafe {
+                glib_ffi::g_free(self.ptr as *mut _);
+            }
+        }
+    }
+}
+
+impl PartialEq for CStringHolder {
+    fn eq(&self, other: &CStringHolder) -> bool {
+        self.slice == other.slice
+    }
+}
+
+impl Eq for CStringHolder {}
+
+impl AsRef<str> for CStringHolder {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl Deref for CStringHolder {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl From<CStringHolder> for String {
+    fn from(holder: CStringHolder) -> Self {
+        String::from(holder.as_str())
+    }
+}
+
+impl From<String> for CStringHolder {
+    fn from(s: String) -> Self {
+        CStringHolder::new(s.into_bytes().as_ptr() as *const c_char)
+    }
+}
+
+impl From<CStringHolder> for Box<str> {
+    fn from(holder: CStringHolder) -> Self {
+        Box::from(holder.as_str())
+    }
+}
+
+impl FromGlibPtrFull<*const c_char> for CStringHolder {
+    unsafe fn from_glib_full(ptr: *const c_char) -> Self {
+        CStringHolder::new(ptr)
+    }
+}
+
+impl FromGlibPtrFull<*mut i8> for CStringHolder {
+    unsafe fn from_glib_full(ptr: *mut i8) -> Self {
+        CStringHolder::new(ptr as *const c_char)
+    }
+}
+
+impl FromGlib<*mut i8> for CStringHolder {
+    fn from_glib(ptr: *mut i8) -> Self {
+        CStringHolder::new(ptr as *const c_char)
+    }
+}
+
+impl FromGlibPtrNone<*const c_char> for CStringHolder {
+    unsafe fn from_glib_none(ptr: *const c_char) -> Self {
+        assert!(!ptr.is_null());
+        CStringHolder::new(ptr)
+    }
+}
+
+impl FromGlibPtrNone<*mut i8> for CStringHolder {
+    unsafe fn from_glib_none(ptr: *mut i8) -> Self {
+        assert!(!ptr.is_null());
+        CStringHolder::new(ptr)
```

Comment:
This would then free the pointer we don't own. We can't have the `from_glib_none()` impls for this type as it owns the pointer

Reply:
Hum the from_glib_none impl is needed by impl_from_glib_container_as_vec_string

Reply:
Hrm, `unimplemented!()` then. This should never be called :)